File: 0odd-even-jump/0odd-even-jump.java
class Solution {
    public int oddEvenJumps(int[] arr) {
        int count = 1;
        boolean[] highJump = new boolean[arr.length], lowJump = new boolean[arr.length];
        highJump[arr.length - 1] = lowJump[arr.length - 1] = true;
        
        TreeMap<Integer, Integer> passedIndexes = new TreeMap();
        passedIndexes.put(arr[arr.length - 1], arr.length - 1);
        
        for (int i = arr.length - 2; i >= 0; i--) {
            Map.Entry<Integer, Integer>
                nextLargest = passedIndexes.ceilingEntry(arr[i]),
                nextSmallest = passedIndexes.floorEntry(arr[i]);
            
            if (nextLargest != null) {
                highJump[i] = lowJump[nextLargest.getValue()];
            } else {
                highJump[i] = false;
            }
            
            if (nextSmallest != null) {
                lowJump[i] = highJump[nextSmallest.getValue()];
            } else {
                lowJump[i] = false;
            }
            
            if (highJump[i]) {
                count++;
            }
            passedIndexes.put(arr[i], i);
        }
        return count;
    }
}

File: 0odd-even-jump/0odd-even-jump.cpp
class Solution {
public:
    int oddEvenJumps(vector<int>& arr) {
        int count = 1;
        std::vector<bool> highJump(arr.size()), lowJump(arr.size());
        highJump[arr.size() - 1] = lowJump[arr.size() - 1] = true;

        std::map<int, int> passedIndexes;
        passedIndexes[arr[arr.size() - 1]] = arr.size() - 1;

        for (int i = arr.size() - 2; i >= 0; i--) {
            auto nextLargest = passedIndexes.lower_bound(arr[i]);
            auto nextSmallest = passedIndexes.upper_bound(arr[i]);

            if (nextLargest != passedIndexes.end()) {
                highJump[i] = lowJump[nextLargest->second];
            } else {
                highJump[i] = false;
            }

            if (nextSmallest != passedIndexes.begin()) {
                lowJump[i] = highJump[(--nextSmallest)->second];
            } else {
                lowJump[i] = false;
            }

            if (highJump[i]) {
                count++;
            }
            passedIndexes[arr[i]] = i;
        }
        return count;
    }
};